Application version handling.

tejlmand opened the pull request:
This PR is inspired by #55814 proposal.

It tries to address the concerns raised in the review.

Main parts of this proposal :
- Zephyr version infrastructure is improved for better reuse, this allows a single infrastructure instead of duplicated files
  This means the same APP_VERSION_x variables similar to KERNEL_VERSION_x are available, providing consistency.
- APP_BUILD_VERSION is supported which includes extra build info through `git describe`, but executed in the application repo.
   For example:
   `#define KERNEL_BUILD_VERSION zephyr-v3.3.0-1081-gb6f97a492717`
   `#define APP_BUILD_VERSION v2.2.0-11-g9e229e382dd2`
- app VERSION file is placed in application folder and follows the semantics of Zephyr kernel VERSION file.
- A dedicated `MCUBOOT_IMGTOOL_SIGN_VERSION` defined, this value defaults to the APP_VERSION_STRING when application is in use, else it defaults to the old `0.0.0+0` string.

Fixes #56605

tejlmand committed:
cmake: generalize VERSION infrastructure for better reuse

Generalize the VERSION and version.h generation so that the same
infrastructure can be reused for generating other version related
header files, such as an application version header.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: generated application version header when VERSION file is present

Allow applications to create a VERSION file in their application folder.

The presence of a VERSION file will create a app_version_h target which
generates an app_version.h file using same infrastructure as the Zephyr
kernel version.h generation.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
west: sign: Add Kconfig with application version

Adds a Kconfig option which controls the version of the application
to use when the image is signed using imgtool.

When an application VERSION file is present, the default value will be
identical to the application version, else it will be 0.0.0+0, but a
project may still decide another value, if it so prefers.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nordicjm reviewed (changes requested):
Some issues with this:

- A dependency issue, modified hello_world to have a sample VERSION file then added `#include "app_version.h"` to main.c and built: `/tmp/aa/zephyr/samples/hello_world/src/main.c:8:10: fatal error: app_version.h: No such file or directory` - had to manually run `ninja app_version_h` to have this file generated
- Version that is generated is not valid for mcuboot, given the below VERSION file, signing then fails: `Error: Invalid value for '-v' / '--version': Invalid version number, should be maj.min.rev+build with later parts optional FATAL ERROR: command exited with status 2: /home/jamie/.local/bin/imgtool sign --version 1.2.3-5 --align 4 --header-size 512 --slot-size 483328 --key /tmp/aa/bootloader/mcuboot/root-rsa-2048.pem /tmp/aa/zephyr/samples/hello_world/_AA/zephyr/zephyr.bin /tmp/aa/zephyr/samples/hello_world/_AA/zephyr/zephyr.signed.bin`
- Changes fail to be picked up - if a project is configured and built, then VERSION is changed and you build again, the old version is used, the old version is also provided to the signing version Kconfig and actually you can't even force the new version to be propagated through manually, running `ninja app_version_h` just gives `ninja: no work to do.`
```
VERSION_MAJOR = 1
VERSION_MINOR = 2
PATCHLEVEL = 3
VERSION_TWEAK = 4
EXTRAVERSION = 5
```

This does not work well.

nordicjm commented:
Also not sure how this would work with something like TF-M whereby there are multiple versions for different components in a single build - maybe @joerchan can see if this would work

tejlmand commented:
> Also not sure how this would work with something like TF-M whereby there are multiple versions for different components in a single build

The application version is not enforced on to TF-M, and regardless of #56141 or #55814 approach it shouldn't.
What both approaches can offer is a default value when a version is active, but still allow users to individually specify TF-M version.

nordicjm commented:
> > Also not sure how this would work with something like TF-M whereby there are multiple versions for different components in a single build
> 
> The application version is not enforced on to TF-M, and regardless of #56141 or #55814 approach it shouldn't. What both approaches can offer is a default value when a version is active, but still allow users to individually specify TF-M version.

Not the version of TF-M itself, the version of the secure core and non-secure core code that is being built, there are currently 2 Kconfigs for this, with the other PR this works fine because each image would get its own Kconfig, with this that is not possible.

> So if you used the same 1.2.3-5 as CONFIG_APPLICATION_VERSION in https://github.com/zephyrproject-rtos/zephyr/pull/55814, then the same would happen.

This generates the wrong version, the other PR does not, the other PR would have to have someone purposely ignore the Kconfig text and enter an invalid value.

nordicjm commented:
And it looks like yours completely changes the kernel version string which breaks existing functionality from what the failing tests show

tejlmand commented:
> And it looks like yours completely changes the kernel version string which breaks existing functionality from what the failing tests show

A `KERNEL` too much, so that it happened to become `KERNEL_BUILD_VERSION` instead of just `BUILD_VERSION`.
Fixed now.

tejlmand commented:
> the old version is also provided to the signing version Kconfig and actually you can't even force the new version to be propagated 

This happens to be an existing bug in Zephyr, just raised an issue here:
https://github.com/zephyrproject-rtos/zephyr/issues/56235

and is exactly why I don't want to have multiple different infrastructures for version handling.
It increases the risk of a dependency issue fixed in one infrastructure is not applied for other infrastructures.

tejlmand commented:
>  there are currently 2 Kconfigs for this, with the other PR this works fine because each image would get its own Kconfig, with this that is not possible.

This PR doesn't remove any existing options, meaning that TF-M still has their existing Kconfig settings, and they should continue to have so, cause how TF-M decides to do versioning is for TF-M project to decide, and Zephyr Kconfig interface to TF-M by providing identical settings.

This means users can continue to specify TF-M secure and non-secure version as they have always done.
What this PR provides, is an ability to automatically have a default based on Zephyr's app version if we wish so.

Same with the imgtool version, users are free to set the version value, and it is not forced to be the same for app and imgtool, although I personally would recommend to have them in sync.

This is exactly why imgtool should have its own Kconfig setting, as mentioned here: https://github.com/zephyrproject-rtos/zephyr/pull/55814#discussion_r1143003499
> if the intention is to be used for img tool, then the setting name should reflect this and the setting should be located with similar img tool setting.

nordicjm commented:
So we go for a universal version but then have exclusions and have different systems that have their own version fields which are completely unrelated? That doesn't fix anything, it just brings confusion and a set of invisible rules.

tejlmand commented:
> A dependency issue, modified hello_world to have a sample VERSION file then added ...

Added a dependency from `zephyr_interface`.
Altough Zephyr itself strictly speaking doesn't use this feature, then the app lib most likely will use it.
As well as any user specific Zephyr modules could want to use this header.
This ensures the `app_version.h` is always generated as part of the build.

> Version that is generated is not valid for mcuboot, given the below VERSION file, signing then fails

Fixed, default version for `MCUBOOT_IMGTOOL_SIGN_VERSION` now follows the `major.minor.patch` provided by the `VERSION` file, else it defaults to `0.0.0+0`.
https://github.com/zephyrproject-rtos/zephyr/blob/b30f828b9de1f0bf0f278f4d8e98f705fd0948cf/Kconfig.zephyr#L861-L864

henrikbrixandersen commented on CMakeLists.txt:
```diff
@@ -544,12 +544,29 @@ add_custom_command(
+      -DVERSION_FILE=${APPLICATION_SOURCE_DIR}/VERSION
+      ${build_version_argument}
+      -P ${ZEPHYR_BASE}/cmake/gen_version_h.cmake
+    DEPENDS ${ZEPHYR_BASE}/VERSION ${git_dependency}
```
Should this be `${APPLICATION_SOURCE_DIR}/VERSION`?

tejlmand replied:
of course it should.
Thanks for spotting

mbolivar-nordic reviewed (dismissed):
This is application facing so it is end user facing. This means documentation is required.

carlescufi commented:
> This is application facing so it is end user facing. This means documentation is required.

Can we split it up and provide the doc in a follow-up PR?

mbolivar-nordic commented:
> Can we split it up and provide the doc in a follow-up PR? (We’ve done this before)

Sure -- @tejlmand?

tejlmand commented:
> > Can we split it up and provide the doc in a follow-up PR? (We’ve done this before)
> 
> Sure -- @tejlmand?

Yep, let's do that.

TomzBench commented:
Docs added yet?  couldn't find

nordicjm commented:
Docs are pending review in https://github.com/zephyrproject-rtos/zephyr/pull/61068

> What happens if their are multiple apps in a same repo, they all get the same git-hash string, with independent version tracking? A "Manifest"/VERSION makes more sense if using git hash i think.

That should be explained in the doc, each application has its own VERSION file which is independent, only one define/variable would remain the same which would be the `git describe` one.

Different versions for different builds of the same application (i.e. with different configurations or for different boards) are impossible and unsupported, however.

marc-hb commented:
> `#define KERNEL_BUILD_VERSION zephyr-v3.3.0-1081-gb6f97a492717`

So the build / version tweak is actually limited to one byte?
- #74357

I'm not sure, afraid I'm getting lost in the details.
